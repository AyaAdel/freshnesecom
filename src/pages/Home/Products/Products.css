.products {
  display: grid;
  gap: 30px 32px;
  grid-template-columns: repeat(3, minmax(269px, 1fr));
}
