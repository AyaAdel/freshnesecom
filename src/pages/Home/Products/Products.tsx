import { useState } from 'react';
import Product from 'components/Product/Product';
import type { AllProducts, ProductItem } from 'types/products';
import { useAppSelector, useAppDispatch } from 'app/hooks';
import type { RootState, AppDispatch } from 'app/store';
import { loadProducts } from 'features/products/productsSlice';
import { getAllProducts } from 'services/products';
import InfiniteScroll from 'react-infinite-scroll-component';

import './Products.css';

export default function Products() {
  const [pageNumber, setPageNumber] = useState<number>(0);
  const [hasMore, setHasMore] = useState(true);

  const dispatch: AppDispatch = useAppDispatch();

  const products: AllProducts = useAppSelector(
    (state: RootState) => state.products
  );

  const pageSize = 9;

  const fetchMoreData = () => {
    if (products.products.length >= products.total) {
      setHasMore(false);
      return;
    }

    setPageNumber((prevPageNumber) => prevPageNumber + 1);

    getAllProducts(pageSize, pageSize * pageNumber + 1).then(
      (response: AllProducts) => {
        dispatch(
          loadProducts({
            ...response,
            products: [...products.products, ...response.products],
          })
        );
      }
    );
  };

  return (
    <>
      {products.products && (
        <InfiniteScroll
          dataLength={products.products.length}
          next={fetchMoreData}
          hasMore={hasMore}
          loader={<h4>Loading...</h4>}
        >
          <div className="products">
            {products.products.map((product: ProductItem) => (
              <Product key={product.id} {...product} />
            ))}
          </div>
        </InfiniteScroll>
      )}
    </>
  );
}
