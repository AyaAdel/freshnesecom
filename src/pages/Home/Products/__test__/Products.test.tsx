import React from 'react';
import renderer from 'react-test-renderer';
import { RenderComponent } from 'utils/test/render';
import Products from '../Products';

function MyComponent() {
  return (
    <RenderComponent>
      <Products />
    </RenderComponent>
  );
}

it('Render Products Correctly', () => {
  const tree = renderer.create(<MyComponent />).toJSON();
  expect(tree).toMatchSnapshot();
});
