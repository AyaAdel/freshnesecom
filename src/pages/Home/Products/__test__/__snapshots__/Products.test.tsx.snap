// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Products Correctly 1`] = `null`;
